app/models/agents/csv_agent.rb

```diff
@@ -0,0 +1,194 @@
+module Agents
+  class CsvAgent < Agent
+    include FormConfigurable
+    include FileHandling
+
+    cannot_be_scheduled!
+
+    def default_options
+      {
+        'mode' => 'parse',
+        'separator' => ',',
+        'use_fields' => '',
+        'output' => 'event_per_row',
+        'with_header' => 'true',
+        'data_path' => '$.data',
+        'data_key' => 'data'
+      }
+    end
+
+    description do
+      <<-MD
+        The `CsvAgent` parses or serializes CSV data. When parsing events are emitted for every row or one event containing all rows.
+
+        `mode` determines if the agent is parsing CSV from incoming event serializes the data of events to CSV.
+
+        ### Universal options
+
+        Specify the `separator` which is used to seperate the fields from each other.
+
+        `data_key` sets in the key which contains the serialized CSV or parsed CSV data in the emitted event.
+
+        ### Parsing
+
+        If `use_fields` is set to a comma seperated string and the CSV file contains field headers the agent will only extract the specified fields.
+
+        `output` determines wheather one event per row is emitted or one event that includes all the rows.
+
+        Set `with_header` to `true` if first line of the CSV file are field names.
+
+        #{receiving_file_handling_agent_description}
+
+        When receiving the CSV data in a regular event use [JSONPath](http://goessner.net/articles/JsonPath/) to select the path in `data_path`. `data_path` is only used when the received event does not contain a 'file pointer'.
+
+        ### Serializing
+
+        If `use_fields` is set to a comma seperated string and the first received event has a object at the specified `data_path` the generated CSV will only include the given fields.
```

Comment:
I suspect this could become an issue later, or at least something we need to keep in mind, since historically I've assumed that events are independent and `receive` taking an array is mostly just an optimization. I think this is the first time it has additional meaning.  Is it possible to do this in a different way and still keep the behavior you want?